src/messaging/ReliableMessageContext.h

```diff
@@ -172,6 +177,13 @@ class ReliableMessageContext
 
         /// When set, ignore session being released, because we are releasing it ourselves.
         kFlagIgnoreSessionRelease = (1u << 10),
+
+        /// When set:
+        ///
+        /// (1) We sent a message that expected a response (hence
+        ///     IsResponseExpected() is true).
+        /// (2) We have received neither a response nor an ack for that message.
+        kFlagWaitingForResponseOrAck = (1u << 11),
```

Comment:
Do we really need this flag? Because once we receive the Message response, we would be practically done with the Exchange and not care about MRP Acks. And there is only one timeout for the Message response(compared to multiple for MRP retries)

Reply:
> we would be practically done with the Exchange

That's not at all true.  Many exchanges have long trains of messages going back and forth.

For the rest, see https://github.com/project-chip/connectedhomeip/pull/29640#discussion_r1350737347

Reply:
Yes, that we do. But we do not have multiple outstanding sent messages. So, an Exchange is a StopAndWait protocol and one and only one message is waiting for a delivery Ack at a given time. 
I might have overlooked the case where there are multiple back-and-forth over a single exchange, but my point was about the single sent outstanding message. It is not a sliding window where a received response is not acking an outstanding sent message(since there is only one outstanding sent message).

Reply:
Sure.  Again, we could treat a non-duplicate response with a piggyback ack as an implicit ack for "whatever the last thing we sent was, even if we have no way to check whether the piggyback is for the right thing", as https://github.com/project-chip/connectedhomeip/pull/29640#discussion_r1350737347 describes.  If you think that would be clearer, we can do that.  We still need two flags to track the "no ack" and "we have no way to check" states, which are not identical.

Reply:
While it is technically true in `our stop-and-wait scenario`, that a received response is actually a true acknowledgment for the message sent earlier(even if all MRP retries have been exhausted), we can ignore the piggybacked Ack in the response(if MRP timeout has happened), but send the response up to the application if the response timeout has not happened and the exchange is still open. For the 2 states, I think WaitingForAck for MRPAck and IsResponseExpected for app response should cover the 2 scenarios, isn't it?

Reply:
> if MRP timeout has happened

This parenthetical is really important: There is nothing tracking whether this has happened right now.  We could have a bit to track _this_, if we want, and then that bit plus the "waiting for ack" bit plus the "response expected" bit would let us detect the states we want.

> For the 2 states

There are more than 2 states.  See https://github.com/project-chip/connectedhomeip/pull/29640#discussion_r1353085415